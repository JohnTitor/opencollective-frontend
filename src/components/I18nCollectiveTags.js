import PropTypes from 'prop-types';
import { defineMessages } from 'react-intl';
import { CollectiveCategory } from '../constants/collectives';
import withIntl from '../lib/withIntl';

const TranslatedTags = defineMessages({
  [CollectiveCategory.ASSOCIATION]: {
    id: 'Tags.ASSOCIATION',
    defaultMessage: 'Association',
  },
  [CollectiveCategory.COLLECTIVE]: {
    id: 'Tags.COLLECTIVE',
    defaultMessage: 'Collective',
  },
  [CollectiveCategory.CONFERENCE]: {
    id: 'Tags.CONFERENCE',
    defaultMessage: 'Conference',
  },
  [CollectiveCategory.COOPERATIVE]: {
    id: 'Tags.COOPERATIVE',
    defaultMessage: 'Cooperative',
  },
  [CollectiveCategory.OPEN_SOURCE]: {
    id: 'Tags.OPEN_SOURCE',
    defaultMessage: 'Open source',
  },
  [CollectiveCategory.MEDIA]: {
    id: 'Tags.MEDIA',
    defaultMessage: 'Media',
  },
  [CollectiveCategory.MEETUP]: {
    id: 'Tags.MEETUP',
    defaultMessage: 'Meetup',
  },
  [CollectiveCategory.MOVEMENT]: {
    id: 'Tags.MOVEMENT',
    defaultMessage: 'Movement',
  },
  [CollectiveCategory.POLITICS]: {
    id: 'Tags.POLITICS',
    defaultMessage: 'Politics',
  },
  [CollectiveCategory.TECH_MEETUP]: {
    id: 'Tags.TECH_MEETUP',
    defaultMessage: 'Tech meetup',
  },
  [CollectiveCategory.US_NONPROFIT]: {
    id: 'Tags.US_NONPROFIT',
    defaultMessage: 'US nonprofit',
  },
});

/** Translates a list of tags */
const I18nCollectiveTags = ({ intl, children, tags, ignoreUntranslated }) => {
  if (typeof tags === 'string') {
    tags = [tags];
  }

  const processedTags = tags.map(tag => {
    if (TranslatedTags[tag]) {
      return { value: intl.formatMessage(TranslatedTags[tag]), isTranslated: true };
    } else {
      return { value: tag, isTranslated: false };
    }
  });

  return children(ignoreUntranslated ? processedTags.filter(t => t.isTranslated) : processedTags);
};

I18nCollectiveTags.propTypes = {
  /** A tag or a list of tags */
  tags: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]).isRequired,
  /** Ignore tags if translation is missing */
  ignoreUntranslated: PropTypes.bool,
  /** A function used to render the tag */
  children: PropTypes.func.isRequired,
  /** @ignore */
  intl: PropTypes.object,
};

I18nCollectiveTags.defaultProps = {
  ignoreUntranslated: false,
  /** Default renderer, will render a string list */
  children: tags => {
    return tags.map((tag, index, translatedTags) => {
      if (index === translatedTags.length - 1) {
        return tag.value;
      } else {
        return `${tag.value}, `;
      }
    });
  },
};

export default withIntl(I18nCollectiveTags);
