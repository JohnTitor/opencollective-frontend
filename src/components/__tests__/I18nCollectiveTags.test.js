import React from 'react';
import I18nCollectiveTags from '../I18nCollectiveTags';
import 'jest-styled-components';
import { snapshot } from '../../../test/snapshot-helpers';

describe('I18nCollectiveTags', () => {
  it('renders a single tag', () => {
    snapshot(<I18nCollectiveTags tags="open source" />);
    snapshot(<I18nCollectiveTags tags="not translated" />);
  });

  it('renders a list of tag', () => {
    snapshot(<I18nCollectiveTags tags={['open source']} />);
    snapshot(<I18nCollectiveTags tags={['not translated', 'open source', 'Meetup']} />);
  });

  it('can ignore untranslated', () => {
    snapshot(<I18nCollectiveTags tags={['not translated', 'open source', 'Meetup']} ignoreUntranslated />);
  });

  it('can provide a custom renderer', () => {
    snapshot(
      <I18nCollectiveTags tags={['HelloWorld', 'Meetup']}>
        {tags =>
          tags.map(({ value, isTranslated }) => (
            <div key={value}>
              {value} is {isTranslated ? 'translated' : 'untranslated'}
            </div>
          ))
        }
      </I18nCollectiveTags>,
    );
  });
});
