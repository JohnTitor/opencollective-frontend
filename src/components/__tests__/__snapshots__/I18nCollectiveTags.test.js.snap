// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I18nCollectiveTags can ignore untranslated 1`] = `
Array [
  "Open source, ",
  "Meetup",
]
`;

exports[`I18nCollectiveTags can provide a custom renderer 1`] = `
Array [
  <div>
    HelloWorld
     is 
    untranslated
  </div>,
  <div>
    Meetup
     is 
    translated
  </div>,
]
`;

exports[`I18nCollectiveTags renders a list of tag 1`] = `"Open source"`;

exports[`I18nCollectiveTags renders a list of tag 2`] = `
Array [
  "not translated, ",
  "Open source, ",
  "Meetup",
]
`;

exports[`I18nCollectiveTags renders a single tag 1`] = `"Open source"`;

exports[`I18nCollectiveTags renders a single tag 2`] = `"not translated"`;
